### PR TITLE
Add "Risk and Confidence HTML" report template

### PR DESCRIPTION
**Pull request by whiteroses**
Signed-off-by: Ira Lun <sammyrosajoe@gmail.com>

As suggested by @psiinon, this is a work-in-progress PR of the report template that I'm submitting for the ZAP Report Competition -- hope this is something like what you are looking for? Will be adding further commits to improve it and add features, including the alert tags as suggested by @psiinon, and will add notes and questions here on various issues that I ran into while working on this -- hope that's ok.

### CODE REVIEW

**Comment by psiinon** [comment]
Just tried this ... and I absolutely LOVE it!
Great work!
<details>
<summary>
Cropped screen shot</summary>

Screenshot 2021-10-01 at 12-52-45 ZAP Scanning Report

</details>

**Comment by whiteroses** [comment]
@psiinon So glad you like it! :) I'd love for it to be included in 2.11.0 -- would really like it to be as good as possible of course, but wouldn't want to delay the release. It's useful as a report template now, but there are a few issues I think you might want to look at before release -- I can write them up over the weekend, and see what other improvements I can make as well while you look at them. Once those issues are resolved (hopefully by around midweek next week), I'm happy for it to be released if you're happy with it :) Would that be ok with you?

File: addOns/reports/src/main/zapHomeFiles/reports/risk-confidence-html/report.html [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+				</div>
+			</section>
+
+			<section data-th-if="${reportData.isIncludeSection('reportParameters')}" id="report-parameters">
+				<h3 data-th-text="#{report.template.section.reportParameters}">Report parameters</h3>
+				<div class="report-parameters--container">
+					<h4 data-th-text="#{report.template.reportParameters.contexts}">Contexts</h4>
+					<th-block data-th-with="includedContexts=${reportData.getContexts()}">
+						<th-block data-th-switch="${includedContexts.size}">
+							<p data-th-case="0">No contexts were selected, so all contexts were included by default.</p>
```
Comment (psiinon) [inline addOns/reports/src/main/zapHomeFiles/reports/risk-confidence-html/report.html:75]: This and some of the other strings should ideally all be i18ned and therefore in the properties file.

**Comment by psiinon** [comment]
Looks great!
You will need to run `./gradlew :addOns:reports:spotlessApply` fir it to pass the build.
It would be good to have all of the strings i18n'ed but thats not a blocker for me...

File: addOns/reports/src/main/zapHomeFiles/reports/risk-confidence-html/report.html (hunk 1 above)
Comment (whiteroses) [reply addOns/reports/src/main/zapHomeFiles/reports/risk-confidence-html/report.html]: Fixed - thanks!

**Comment by psiinon** [comment]
The build is still failing - you should be able to fix that using `./gradlew :addOns:reports:spotlessApply` :)

**Comment by psiinon** [comment]
Oh, and you'll also need to remove the [WIP] from the title (or we can do that;)

**Comment by whiteroses** [comment]
@psiinon re: the active scan rules issue we discussed yesterday, I'm able to reliably reproduce the issue on my machine, and it seems to be something to do with session/database persistence.

To reproduce:

If ZAP is not configured to persist session from the start:
1) Run and complete an active scan (I tested this on the OWASP Juice Shop).
2) Generate Report with "Traditional HTML Report with requests and responses" template and "Passing Rules" section ticked (that's the only section that matters for this). Both Active and Passive rules are found in the Passing Rules table (as I think you found yesterday).
3) Choose File > Persist Session..., then generate the report again. Now only Passive rules can be found in the Passing Rules table. (Don't know if this is related, but when I chose Persist Session with the Active Scan tab open, the Active Scan tab seemed to clear itself, which seemed like surprising behaviour.)
4) Exit and restart ZAP, then load the persisted session, and generate report. Again, only Passive rules can be found in the Passing Rules table.

If ZAP is set up to persist session from the start, I'm able to reproduce the issue by exiting and restarting ZAP, loading the persisted session, and generating the report (step 4 above).

I think this has to do with `reportData.reportObjects.get('activeScanData')` returning `null` (as in the traditional-html-plus template). Would you mind testing this and letting me know if it reproduces for you on your machine?

With `./gradlew :addOns:reports:spotlessApply`, would it be ok if I run it and commit the changes when you decide you want to merge? The changes make the code somewhat more difficult to work with, and I will have to run it again at the end as well to be sure. Would that be ok?

**Comment by whiteroses** [comment]
I'd still like to work on this a little more, but as discussed, if you ever decide you want to merge or need it for release, please let me know and I'll do what's necessary to wrap it up, run `./gradlew :addOns:reports:spotlessApply` and commit the changes, and remove the [WIP]. Are you looking to merge now, or would you just like me to run `./gradlew :addOns:reports:spotlessApply` so that you can see that the build passes?

**Comment by thc202** [comment]
> re: the active scan rules issue we discussed yesterday,

That's the expected behaviour the scans' state is not persisted.

**Comment by psiinon** [comment]
That problem is a known restriction - basically the ascan data is held in memory and that gets lots if you change the session, even just persisting it.
Fixing that will be non trivial :/
We cant merge until the build passed and WIP is removed - it would be really good it you could do that. 
You'll still be able to make changes, but they may not get into the 2.11 release, depending on when that is started ... but it is imminent ;)

**Comment by whiteroses** [comment]
Ok I understand; it's just that it leads to really unexpected behaviour for a user (have I missed it in the documentation?) Does that mean that anything that relies on ascan data *must* be done before ever exiting ZAP or choosing to persist the session? And that e.g. the Passing Rules section in the traditional-html-plus report (and all the other templates with the same section) would not work as expected if the user exits ZAP or chooses to persist the session? I understand that the fix is non-trivial -- perhaps that means I should take that into account and limit any reliance on the ascan data in my template as well.

With the imminent release -- would wrapping it up by tomorrow or Thursday be ok and make it in time for the 2.11 release, or would you prefer that I make it ready for merging today?

**Comment by psiinon** [comment]
Yeah, we should probably make that clearer in the docs.
Could you make it ready for merging asap, ie today?
You'll be able to submit new PRs up until the point we release, and after that although then we'll need to re-release the reporting add-on, which is not a big deal tbh.

**Comment by whiteroses** [comment]
Ok, I'll get it ready by the end of today :)

**Comment by whiteroses** [comment]
It seems Spotless changed `<th-block data-th-each="confidenceCode, iterStat: ${includedConfidenceCodes}"><th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block><span class="confidence-level" data-th-text="${helper.getConfidenceString(confidenceCode)}">confidence level</span></th-block>` (intentionally one consecutive line, because of [this](https://github.com/thymeleaf/thymeleaf/issues/586)) into


```

                                                                                                                                   <th-block                                           
                                                                data-th-each="confidenceCode, iterStat: ${includedConfidenceCodes}">                          
                                                        <th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block>                                 
                                                        <span class="confidence-level"                                                                        
                                                                data-th-text="${helper.getConfidenceString(confidenceCode)}">confidence                       
                                                                level</span></th-block>

```

so there is now a space in front of the commas, e.g. <pre><code>Included: High , Medium , Low , Informational</code></pre>

I guess I could probably get around it by concatenating the strings with the HTML tags and using data-th-utext? But is that what you'd like me to do, or is there maybe a switch or option to stop Spotless doing this?

Other than that, I think this is ready for you :)

**Comment by whiteroses** [comment]
Noticed another similar problem: Spotless turned `<pre data-th-case="true"><code data-th-text="${userObject.message.requestHeader}">Request line and header section</code></pre>` into 

```html
<pre data-th-case="true">
	<code data-th-text="${userObject.message.requestHeader}">Request line and header section</code>
</pre>
```

https://github.com/zaproxy/zap-extensions/blob/31de5d2efead988cdb9cfd8b9f4259ca28adfbcb/addOns/reports/src/main/zapHomeFiles/reports/risk-confidence-html/report.html#L649

If you generate a report with the template (with requests and responses included), you can see the whitespace messing up the layout of the request and response request/status lines and bodies.

**Comment by psiinon** [comment]
We can disable to formatting check for specific files, and I'm fine doing that in this case.
Is it ready apart from that?
We're planning on doing the release today ;)

**Review by psiinon — APPROVED** [review]

**Review by kingthorin — APPROVED** [review]

**Comment by whiteroses** [comment]
Sorry I was only using those lines as an example -- there are a few more...

```html
diff --git a/addOns/reports/src/main/zapHomeFiles/reports/risk-confidence-html/report.html b/addOns/reports/src/main/zapHomeFiles/reports/risk-confidence-html/report.html
index 4f9e838fc..e8d60a9d3 100644
--- a/addOns/reports/src/main/zapHomeFiles/reports/risk-confidence-html/report.html
+++ b/addOns/reports/src/main/zapHomeFiles/reports/risk-confidence-html/report.html
@@ -173,9 +173,7 @@
                                                        data-th-case="0"
                                                        data-th-text="#{report.template.reportParameters.none}">None</span>
                                                <span data-th-case="*" class="included-risk-codes"> <th-block
-                                                               data-th-each="riskCode, iterStat: ${includedRiskCodes}">
-                                                       <th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block>
-                                                       <span class="risk-level"
+                                                               data-th-each="riskCode, iterStat: ${includedRiskCodes}"><th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block><span class="risk-level"
                                                                data-th-text="${helper.getRiskString(riskCode)}">risk
                                                                level</span></th-block>
                                                </span> </th-block>
@@ -186,9 +184,7 @@
                                                        data-th-case="${MAX_RISK_CODE + 1}"
                                                        data-th-text="#{report.template.reportParameters.none}">None</span>
                                                <span data-th-case="*" class="included-risk-codes"> <th-block
-                                                               data-th-each="riskCode, iterStat: ${#numbers.sequence(MAX_RISK_CODE, 0, -1).{? #this not in #includedRiskCodes}}">
-                                                       <th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block>
-                                                       <span class="risk-level"
+                                                               data-th-each="riskCode, iterStat: ${#numbers.sequence(MAX_RISK_CODE, 0, -1).{? #this not in #includedRiskCodes}}"><th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block><span class="risk-level"
                                                                data-th-text="${helper.getRiskString(riskCode)}">risk
                                                                level</span></th-block>
                                                </span> </th-block>
@@ -203,9 +199,7 @@
                                                <span data-th-case="0"
                                                        data-th-text="#{report.template.reportParameters.none}">None</span>
                                                <span data-th-case="*" class="included-confidence-codes"> <th-block
-                                                               data-th-each="confidenceCode, iterStat: ${includedConfidenceCodes}">
-                                                       <th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block>
-                                                       <span class="confidence-level"
+                                                               data-th-each="confidenceCode, iterStat: ${includedConfidenceCodes}"><th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block><span class="confidence-level"
                                                                data-th-text="${helper.getConfidenceString(confidenceCode)}">confidence
                                                                level</span></th-block>
                                                </span> </th-block>
@@ -216,9 +210,7 @@
                                                <span data-th-case="${MAX_CONFIDENCE_CODE + 1}"
                                                        data-th-text="#{report.template.reportParameters.none}">None</span>
                                                <span data-th-case="*" class="included-confidence-codes"> <th-block
-                                                               data-th-each="confidenceCode, iterStat: ${#numbers.sequence(MAX_CONFIDENCE_CODE, 0, -1).{? #this not in #includedConfidenceCodes}}">
-                                                       <th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block>
-                                                       <span class="confidence-level"
+                                                               data-th-each="confidenceCode, iterStat: ${#numbers.sequence(MAX_CONFIDENCE_CODE, 0, -1).{? #this not in #includedConfidenceCodes}}"><th-block data-th-text="${iterStat.index == 0? '': ', '}"></th-block><span class="confidence-level"
                                                                data-th-text="${helper.getConfidenceString(confidenceCode)}">confidence
                                                                level</span></th-block>
                                                </span> </th-block>
@@ -646,9 +638,7 @@
                                        line and header section (x bytes)</summary>
                                <th-block
                                        data-th-switch="${reportData.isIncludeSection('requestHeader')}">
-                               <pre data-th-case="true">
-                                       <code data-th-text="${userObject.message.requestHeader}">Request line and header section</code>
-                               </pre>
+                               <pre data-th-case="true"><code data-th-text="${userObject.message.requestHeader}">Request line and header section</code></pre>
                                <p data-th-case="false"
                                        data-th-text="#{report.template.alertsTable.excludedFromReport}">[Excluded
                                        from report]</p>
@@ -661,9 +651,7 @@
                                        body (x bytes)</summary>
                                <th-block
                                        data-th-switch="${reportData.isIncludeSection('requestBody')}">
-                               <pre data-th-case="true">
-                                       <code data-th-text="${userObject.message.requestBody}">Request body</code>
-                               </pre>
+                               <pre data-th-case="true"><code data-th-text="${userObject.message.requestBody}">Request body</code></pre>
                                <p data-th-case="false"
                                        data-th-text="#{report.template.alertsTable.excludedFromReport}">[Excluded
                                        from report]</p>
@@ -680,9 +668,7 @@
                                        line and header section (x bytes)</summary>
                                <th-block
                                        data-th-switch="${reportData.isIncludeSection('responseHeader')}">
-                               <pre data-th-case="true">
-                                       <code data-th-text="${userObject.message.responseHeader}">Status line and header section</code>
-                               </pre>
+                               <pre data-th-case="true"><code data-th-text="${userObject.message.responseHeader}">Status line and header section</code></pre>
                                <p data-th-case="false"
                                        data-th-text="#{report.template.alertsTable.excludedFromReport}">[Excluded
                                        from report]</p>
@@ -695,9 +681,7 @@
                                        body (x bytes)</summary>
                                <th-block
                                        data-th-switch="${reportData.isIncludeSection('responseBody')}">
-                               <pre data-th-case="true">
-                                       <code data-th-text="${userObject.message.responseBody}">Response body</code>
-                               </pre>
+                               <pre data-th-case="true"><code data-th-text="${userObject.message.responseBody}">Response body</code></pre>
                                <p data-th-case="false"
                                        data-th-text="#{report.template.alertsTable.excludedFromReport}">[Excluded
                                        from report]</p>
@@ -707,21 +691,15 @@
        <tr data-th-unless="${userObject.param.isEmpty()}">
                <th scope="row"
                        data-th-text="#{report.template.alertsTable.parameter}">Parameter</th>
-               <td><pre>
-                               <code data-th-text="${userObject.param}">Parameter</code>
-                       </pre></td>
+               <td><pre><code data-th-text="${userObject.param}">Parameter</code></pre></td>
        </tr>
        <tr data-th-unless="${userObject.attack.isEmpty()}">
                <th scope="row" data-th-text="#{report.template.alertsTable.attack}">Attack</th>
-               <td><pre>
-                               <code data-th-text="${userObject.attack}">Attack</code>
-                       </pre></td>
+               <td><pre><code data-th-text="${userObject.attack}">Attack</code></pre></td>
        </tr>
        <tr data-th-unless="${userObject.evidence.isEmpty()}">
                <th scope="row" data-th-text="#{report.template.alertsTable.evidence}">Evidence</th>
-               <td><pre>
-                               <code data-th-text="${userObject.evidence}">Evidence</code>
-                       </pre></td>
+               <td><pre><code data-th-text="${userObject.evidence}">Evidence</code></pre></td>
        </tr>
        <tr data-th-unless="${userObject.solution.isEmpty()}">
                <th scope="row" data-th-text="#{report.template.alertsTable.solution}">Solution</th>
```

Hope that helps?

**Comment by psiinon** [comment]
Many thanks!
I'll work on a PR to fix the problems you mentioned and to disable the spotless check...

**Comment by whiteroses** [comment]
Award received, thanks! :)

**Pull request merged**